src/images/kafka.rs

```diff
@@ -0,0 +1,124 @@
+use crate::{core::WaitFor, Image};
+use std::{collections::HashMap, path::PathBuf};
+
+const CONTAINER_IDENTIFIER: &str = "confluentinc/cp-kafka";
+const DEFAULT_TAG: &str = "6.1.1";
+
+pub const KAFKA_PORT: u16 = 9093;
+const ZOOKEEPER_PORT: u16 = 2181;
+
+#[derive(Clone, Debug)]
+pub struct KafkaArgs {
+    pub bash_pathname: PathBuf,
+    pub options: Vec<String>,
+    pub commands: String,
+}
+
+impl Default for KafkaArgs {
+    fn default() -> Self {
+        let starter_commands = format!(
+            "echo 'clientPort={}' > zookeeper.properties; \
+echo 'dataDir=/var/lib/zookeeper/data' >> zookeeper.properties; \
+echo 'dataLogDir=/var/lib/zookeeper/log' >> zookeeper.properties; \
+zookeeper-server-start zookeeper.properties & \
+export KAFKA_ZOOKEEPER_CONNECT='localhost:{}'; \
+export KAFKA_ADVERTISED_LISTENERS='PLAINTEXT://localhost:{},BROKER://localhost:9092'; \
+. /etc/confluent/docker/bash-config && \
+/etc/confluent/docker/configure && \
+/etc/confluent/docker/launch",
+            ZOOKEEPER_PORT, ZOOKEEPER_PORT, KAFKA_PORT
+        );
+
+        Self {
+            bash_pathname: PathBuf::from("/bin/bash"),
+            options: vec!["-c".to_owned()],
+            commands: starter_commands,
+        }
+    }
+}
+
+impl IntoIterator for KafkaArgs {
+    type Item = String;
+    type IntoIter = ::std::vec::IntoIter<String>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        let mut args = vec![self.bash_pathname.into_os_string().into_string().unwrap()];
+        args.extend(self.options.into_iter());
+        args.push(self.commands);
+        args.into_iter()
+    }
+}
```

Comment:
This looks fairly inconsistent to me:

The fields are public and therefore configurable. However, from the way they are populated within `default`, it doesn't seem like a user would get a working container if they override any this.

As a result, it looks like we don't need any of these fields and can just inline everything into the `IntoIterator` implementation. Generally, the way this is meant to be used is:

- the `XYZArgs` struct has strongly typed fields that - using the type system - model the full range of valid options.
- the `IntoIterator` implementation uses those fields and converts them into whatever command string necessary to run the container

It looks like none of what is stored within `KafkaArgs` makes sense to be configurable so concluding from that, all of this should just be inlined into the `IntoIterator` implementation and `KafkaArgs` becomes a zero-sized struct.

Reply:
Thanks for the clarification. Now I see what you mean. Since they are rather mandatory & fixed values, hard-coding them in `IntoIterator` makes perfect sense to me. Addressed in 81be19a.